File: Samples/NineSlice/C#/Src/NineSliceImage.cs
#if NOESIS
using Noesis;
#else
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;
#endif
using System;

namespace NineSlice
{
    /// <summary>
    /// Nine slice image control
    /// </summary>
    public class NineSliceImage : ContentControl
    {
        public ImageSource ImageSource
        {
            get { return (ImageSource)GetValue(ImageSourceProperty); }
            set { SetValue(ImageSourceProperty, value); }
        }

        public static readonly DependencyProperty ImageSourceProperty = DependencyProperty.Register(
            "ImageSource", typeof(ImageSource), typeof(NineSliceImage),
            new PropertyMetadata(null, OnSlicesChanged));

        public Thickness Slices
        {
            get { return (Thickness)GetValue(SlicesProperty); }
            set { SetValue(SlicesProperty, value); }
        }

        public static readonly DependencyProperty SlicesProperty = DependencyProperty.Register(
            "Slices", typeof(Thickness), typeof(NineSliceImage),
            new PropertyMetadata(new Thickness(), OnSlicesChanged));

        #region 9-grid slices
        public Int32Rect TopLeftViewbox
        {
            get { return (Int32Rect)GetValue(TopLeftViewboxProperty); }
            set { SetValue(TopLeftViewboxProperty, value); }
        }

        public static readonly DependencyProperty TopLeftViewboxProperty = DependencyProperty.Register(
            "TopLeftViewbox", typeof(Int32Rect), typeof(NineSliceImage),
            new PropertyMetadata(new Int32Rect()));

        public Int32Rect TopCenterViewbox
        {
            get { return (Int32Rect)GetValue(TopCenterViewboxProperty); }
            set { SetValue(TopCenterViewboxProperty, value); }
        }

        public static readonly DependencyProperty TopCenterViewboxProperty = DependencyProperty.Register(
            "TopCenterViewbox", typeof(Int32Rect), typeof(NineSliceImage),
            new PropertyMetadata(new Int32Rect()));

        public Int32Rect TopRightViewbox
        {
            get { return (Int32Rect)GetValue(TopRightViewboxProperty); }
            set { SetValue(TopRightViewboxProperty, value); }
        }

        public static readonly DependencyProperty TopRightViewboxProperty = DependencyProperty.Register(
            "TopRightViewbox", typeof(Int32Rect), typeof(NineSliceImage),
            new PropertyMetadata(new Int32Rect()));

        public Int32Rect CenterLeftViewbox
        {
            get { return (Int32Rect)GetValue(CenterLeftViewboxProperty); }
            set { SetValue(CenterLeftViewboxProperty, value); }
        }

        public static readonly DependencyProperty CenterLeftViewboxProperty = DependencyProperty.Register(
            "CenterLeftViewbox", typeof(Int32Rect), typeof(NineSliceImage),
            new PropertyMetadata(new Int32Rect()));

        public Int32Rect CenterViewbox
        {
            get { return (Int32Rect)GetValue(CenterViewboxProperty); }
            set { SetValue(CenterViewboxProperty, value); }
        }

        public static readonly DependencyProperty CenterViewboxProperty = DependencyProperty.Register(
            "CenterViewbox", typeof(Int32Rect), typeof(NineSliceImage),
            new PropertyMetadata(new Int32Rect()));

        public Int32Rect CenterRightViewbox
        {
            get { return (Int32Rect)GetValue(CenterRightViewboxProperty); }
            set { SetValue(CenterRightViewboxProperty, value); }
        }

        public static readonly DependencyProperty CenterRightViewboxProperty = DependencyProperty.Register(
            "CenterRightViewbox", typeof(Int32Rect), typeof(NineSliceImage),
            new PropertyMetadata(new Int32Rect()));

        public Int32Rect BottomLeftViewbox
        {
            get { return (Int32Rect)GetValue(BottomLeftViewboxProperty); }
            set { SetValue(BottomLeftViewboxProperty, value); }
        }

        public static readonly DependencyProperty BottomLeftViewboxProperty = DependencyProperty.Register(
            "BottomLeftViewbox", typeof(Int32Rect), typeof(NineSliceImage),
            new PropertyMetadata(new Int32Rect()));

        public Int32Rect BottomCenterViewbox
        {
            get { return (Int32Rect)GetValue(BottomCenterViewboxProperty); }
            set { SetValue(BottomCenterViewboxProperty, value); }
        }

        public static readonly DependencyProperty BottomCenterViewboxProperty = DependencyProperty.Register(
            "BottomCenterViewbox", typeof(Int32Rect), typeof(NineSliceImage),
            new PropertyMetadata(new Int32Rect()));

        public Int32Rect BottomRightViewbox
        {
            get { return (Int32Rect)GetValue(BottomRightViewboxProperty); }
            set { SetValue(BottomRightViewboxProperty, value); }
        }

        public static readonly DependencyProperty BottomRightViewboxProperty = DependencyProperty.Register(
            "BottomRightViewbox", typeof(Int32Rect), typeof(NineSliceImage),
            new PropertyMetadata(new Int32Rect()));
        #endregion

        private static void OnSlicesChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
        {
            NineSliceImage image = (NineSliceImage)d;

            float width = 0.0f;
            float height = 0.0f;

            if (image.ImageSource != null)
            {
                width = (float)image.ImageSource.Width;
                height = (float)image.ImageSource.Height;
            }

            int wl = (int)Math.Min(image.Slices.Left, width);
            int wc = (int)Math.Max(width - image.Slices.Left - image.Slices.Right, 0);
            int wr = (int)Math.Min(image.Slices.Right, width);
            int ht = (int)Math.Min(image.Slices.Top, height);
            int hc = (int)Math.Max(height - image.Slices.Top - image.Slices.Bottom, 0);
            int hb = (int)Math.Min(image.Slices.Bottom, height);

            image.TopLeftViewbox = new Int32Rect(0, 0, wl, ht);
            image.TopCenterViewbox = new Int32Rect(wl, 0, wc, ht);
            image.TopRightViewbox = new Int32Rect(wl + wc, 0, wr, ht);

            image.CenterLeftViewbox = new Int32Rect(0, ht, wl, hc);
            image.CenterViewbox = new Int32Rect(wl, ht, wc, hc);
            image.CenterRightViewbox = new Int32Rect(wl + wc, ht, wr, hc);

            image.BottomLeftViewbox = new Int32Rect(0, ht + hc, wl, hb);
            image.BottomCenterViewbox = new Int32Rect(wl, ht + hc, wc, hb);
            image.BottomRightViewbox = new Int32Rect(wl + wc, ht + hc, wr, hb);
        }
    }
}

File: Samples/NineSlice/C++/Src/NineSliceImage.cpp
////////////////////////////////////////////////////////////////////////////////////////////////////
// NoesisGUI - http://www.noesisengine.com
// Copyright (c) 2013 Noesis Technologies S.L. All Rights Reserved.
////////////////////////////////////////////////////////////////////////////////////////////////////


#include "NineSliceImage.h"

#include <NsCore/ReflectionImplement.h>
#include <NsGui/UIElementData.h>
#include <NsGui/PropertyMetadata.h>
#include <NsGui/ImageSource.h>
#include <NsDrawing/Thickness.h>


using namespace Noesis;
using namespace NineSlice;


////////////////////////////////////////////////////////////////////////////////////////////////////
ImageSource* NineSliceImage::GetImageSource() const
{
    return GetValue<Ptr<ImageSource>>(ImageSourceProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void NineSliceImage::SetImageSource(ImageSource* value)
{
    SetValue<Ptr<ImageSource>>(ImageSourceProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
const Thickness& NineSliceImage::GetSlices() const
{
    return GetValue<Thickness>(SlicesProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void NineSliceImage::SetSlices(const Thickness& value)
{
    SetValue<Thickness>(SlicesProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
static void OnSlicesChanged(DependencyObject* d, const DependencyPropertyChangedEventArgs&)
{
    NineSliceImage* image = (NineSliceImage*)d;

    float width = 0;
    float height = 0;

    ImageSource* imageSource = image->GetImageSource();
    if (imageSource != nullptr)
    {
        width = imageSource->GetWidth();
        height = imageSource->GetHeight();
    }

    const Thickness& slices = image->GetSlices();

    float wl = Min(slices.left, width);
    float wc = Max(width - slices.left - slices.right, 0.0f);
    float wr = Min(slices.right, width);
    float ht = Min(slices.top, height);
    float hc = Max(height - slices.top - slices.bottom, 0.0f);
    float hb = Min(slices.bottom, height);

    image->SetValue<Rect>(NineSliceImage::TopLeftViewboxProperty, Rect(0.0f, 0.0f, wl, ht));
    image->SetValue<Rect>(NineSliceImage::TopCenterViewboxProperty, Rect(wl, 0.0f, wl + wc, ht));
    image->SetValue<Rect>(NineSliceImage::TopRightViewboxProperty, Rect(wl + wc, 0.0f, wl + wc + wr, ht));

    image->SetValue<Rect>(NineSliceImage::CenterLeftViewboxProperty, Rect(0.0f, ht, wl, ht + hc));
    image->SetValue<Rect>(NineSliceImage::CenterViewboxProperty, Rect(wl, ht, wl + wc, ht + hc));
    image->SetValue<Rect>(NineSliceImage::CenterRightViewboxProperty, Rect(wl + wc, ht, wl + wc + wr, ht + hc));

    image->SetValue<Rect>(NineSliceImage::BottomLeftViewboxProperty, Rect(0.0f, ht + hc, wl, ht + hc + hb));
    image->SetValue<Rect>(NineSliceImage::BottomCenterViewboxProperty, Rect(wl, ht + hc, wl + wc, ht + hc + hb));
    image->SetValue<Rect>(NineSliceImage::BottomRightViewboxProperty, Rect(wl + wc, ht + hc, wl + wc + wr, ht + hc + hb));
}

////////////////////////////////////////////////////////////////////////////////////////////////////
NS_BEGIN_COLD_REGION

NS_IMPLEMENT_REFLECTION(NineSliceImage, "NineSlice.NineSliceImage")
{
    UIElementData* data = NsMeta<UIElementData>(TypeOf<SelfClass>());

    data->RegisterProperty<Ptr<ImageSource>>(ImageSourceProperty, "ImageSource",
        PropertyMetadata::Create(Ptr<ImageSource>(), OnSlicesChanged));
    data->RegisterProperty<Thickness>(SlicesProperty, "Slices",
        PropertyMetadata::Create(Thickness(), OnSlicesChanged));

    data->RegisterProperty<Rect>(TopLeftViewboxProperty, "TopLeftViewbox",
        PropertyMetadata::Create(Rect()));
    data->RegisterProperty<Rect>(TopCenterViewboxProperty, "TopCenterViewbox",
        PropertyMetadata::Create(Rect()));
    data->RegisterProperty<Rect>(TopRightViewboxProperty, "TopRightViewbox",
        PropertyMetadata::Create(Rect()));
    data->RegisterProperty<Rect>(CenterLeftViewboxProperty, "CenterLeftViewbox",
        PropertyMetadata::Create(Rect()));
    data->RegisterProperty<Rect>(CenterViewboxProperty, "CenterViewbox",
        PropertyMetadata::Create(Rect()));
    data->RegisterProperty<Rect>(CenterRightViewboxProperty, "CenterRightViewbox",
        PropertyMetadata::Create(Rect()));
    data->RegisterProperty<Rect>(BottomLeftViewboxProperty, "BottomLeftViewbox",
        PropertyMetadata::Create(Rect()));
    data->RegisterProperty<Rect>(BottomCenterViewboxProperty, "BottomCenterViewbox",
        PropertyMetadata::Create(Rect()));
    data->RegisterProperty<Rect>(BottomRightViewboxProperty, "BottomRightViewbox",
        PropertyMetadata::Create(Rect()));
}

////////////////////////////////////////////////////////////////////////////////////////////////////
const DependencyProperty* NineSliceImage::ImageSourceProperty;
const DependencyProperty* NineSliceImage::SlicesProperty;
const DependencyProperty* NineSliceImage::TopLeftViewboxProperty;
const DependencyProperty* NineSliceImage::TopCenterViewboxProperty;
const DependencyProperty* NineSliceImage::TopRightViewboxProperty;
const DependencyProperty* NineSliceImage::CenterLeftViewboxProperty;
const DependencyProperty* NineSliceImage::CenterViewboxProperty;
const DependencyProperty* NineSliceImage::CenterRightViewboxProperty;
const DependencyProperty* NineSliceImage::BottomLeftViewboxProperty;
const DependencyProperty* NineSliceImage::BottomCenterViewboxProperty;
const DependencyProperty* NineSliceImage::BottomRightViewboxProperty;